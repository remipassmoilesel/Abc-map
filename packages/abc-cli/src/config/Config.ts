/*
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 *
 *
 */

import * as path from 'path';

export class Config {
  public getProjectRoot(): string {
    return path.resolve(__dirname, '..', '..', '..', '..');
  }

  public getServerRoot(): string {
    return path.resolve(this.getProjectRoot(), 'packages/server');
  }

  public getServerPublicRoot(): string {
    return path.resolve(this.getServerRoot(), 'public');
  }

  public getCliRoot(): string {
    return path.resolve(this.getProjectRoot(), 'packages/abc-cli');
  }

  public getDevServicesRoot(): string {
    return path.resolve(this.getProjectRoot(), 'packages/infrastructure/dev-services');
  }

  public getE2eRoot(): string {
    return path.resolve(this.getProjectRoot(), 'packages/e2e-tests');
  }

  public getPerformanceTestsRoot(): string {
    return path.resolve(this.getProjectRoot(), 'packages/performance-tests');
  }

  public getUserDocRoot(): string {
    return path.resolve(this.getProjectRoot(), 'packages/user-documentation');
  }

  public getFrontendUrl(): string {
    return 'http://127.0.0.1:10082/';
  }

  public getBackendUrl(): string {
    return 'http://127.0.0.1:10082/api/health/';
  }

  public getChartRoot(): string {
    return path.resolve(this.getProjectRoot(), 'packages/infrastructure/helm-chart/abc-map');
  }
}
