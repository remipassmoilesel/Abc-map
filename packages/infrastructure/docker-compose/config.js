module.exports = {
  environmentName: 'production',
  externalUrl: 'http://localhost:10082/',
  server: {
    host: '0.0.0.0',
    port: 10082,
    log: {
      requests: true,
      errors: true,
      warnings: false,
    },
    globalRateLimit: {
      max: 1000,
      timeWindow: '30m',
    },
    authenticationRateLimit: {
      max: 200,
      timeWindow: '30m',
    },
  },
  project: {
    maxPerUser: 10,
  },
  database: {
    url: 'mongodb://abc-mongodb:27017',
    username: 'CHANGEME',
    password: 'CHANGEME',
  },
  jwt: {
    algorithm: 'HS512',
  },
  authentication: {
    secret: 'CHANGEME',
    tokenExpiresIn: '7d',
    passwordLostExpiresIn: '30min',
  },
  registration: {
    passwordSalt: 'CHANGEME',
    secret: 'CHANGEME',
    confirmationExpiresIn: '24h',
  },
  smtp: {
    from: 'no-reply@abc-map.fr',
    host: 'abc-smtp',
    port: 25,
  },
  datastore: {
    path: '/srv/datastore',
  },
  legalMentions: `
            <div class="p-2">
                <h1>CHANGEME</h1>
            </div>
            `,
};
