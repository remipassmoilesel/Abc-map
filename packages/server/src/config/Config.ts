/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Algorithm } from 'jsonwebtoken';

export const LOCAL_ENVIRONMENT = 'local';
export const TEST_ENVIRONMENT = 'test';
export const STAGING_ENVIRONMENT = 'staging';

export interface ConfigInput {
  environmentName: string;
  development?: DevelopmentDataConfig;
  // External URL is the url that will be used by users. It must contain protocol and full domain. E.g: 'https://abc-map.fr'.
  externalUrl: string;
  server: ServerConfig;
  project: ProjectConfig;
  database: DatabaseConfig;
  jwt: JwtConfig;
  authentication: AuthenticationConfig;
  registration: RegistrationConfig;
  smtp: SmtpConfig;
  datastore: DatastoreConfig;
  legalMentions: string;
  webapp?: WebappConfig;
  userDocumentation?: UserDocConfig;
}

/**
 * This config is used in application
 */
export interface Config extends ConfigInput {
  webappPath: string;
  userDocumentationPath: string;
}

export interface DevelopmentDataConfig {
  // If set, data will be generated in order to facilitate development (users, projects, ...)
  generateData?: {
    users: number;
    projectsPerUser: number;
  };
  // If true, mails will not be sent but persisted on disk
  persistEmails?: boolean;
}

export interface ServerConfig {
  host: string;
  port: number;
  log: {
    requests: boolean;
    errors: boolean;
    warnings: boolean;
  };
  globalRateLimit: {
    max: number;
    timeWindow: string;
  };
  authenticationRateLimit: {
    max: number;
    timeWindow: string;
  };
}

export interface ProjectConfig {
  maxPerUser: number;
}

export interface DatabaseConfig {
  url: string;
  databaseName?: string;
  username: string;
  password: string;
}

export interface JwtConfig {
  algorithm: Algorithm;
}

export interface AuthenticationConfig {
  secret: string;
  tokenExpiresIn: string;
  passwordLostExpiresIn: string;
}

export interface RegistrationConfig {
  passwordSalt: string;
  secret: string;
  confirmationExpiresIn: string;
}

export interface SmtpConfig {
  from: string;
  host: string;
  port: number;
  auth?: {
    user: string;
    pass: string;
  };
}

export interface DatastoreConfig {
  path: string;
}

export interface WebappConfig {
  appendToBody?: string;
}

export interface UserDocConfig {
  appendToBody?: string;
}
