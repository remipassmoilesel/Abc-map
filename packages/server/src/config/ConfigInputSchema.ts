/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { JSONSchemaType } from 'ajv';
import { ConfigInput } from './Config';

export const ConfigInputSchema: JSONSchemaType<ConfigInput> = {
  type: 'object',
  required: ['environmentName', 'externalUrl', 'server', 'project', 'database', 'jwt', 'authentication', 'registration', 'smtp', 'datastore', 'legalMentions'],
  additionalProperties: false,
  properties: {
    environmentName: { type: 'string' },
    externalUrl: { type: 'string' },
    server: {
      type: 'object',
      required: ['host', 'port', 'log', 'globalRateLimit', 'authenticationRateLimit'],
      properties: {
        host: { type: 'string' },
        port: { type: 'number' },
        log: {
          type: 'object',
          required: ['requests', 'errors', 'warnings'],
          properties: {
            requests: { type: 'boolean' },
            errors: { type: 'boolean' },
            warnings: { type: 'boolean' },
          },
        },
        globalRateLimit: {
          type: 'object',
          required: ['max', 'timeWindow'],
          properties: {
            max: { type: 'number' },
            timeWindow: { type: 'string' },
          },
        },
        authenticationRateLimit: {
          type: 'object',
          required: ['max', 'timeWindow'],
          properties: {
            max: { type: 'number' },
            timeWindow: { type: 'string' },
          },
        },
      },
    },
    project: {
      type: 'object',
      required: ['maxPerUser'],
      properties: {
        maxPerUser: { type: 'number' },
      },
    },
    database: {
      type: 'object',
      required: ['url', 'username', 'password'],
      properties: {
        url: { type: 'string' },
        databaseName: { type: 'string', nullable: true },
        username: { type: 'string' },
        password: { type: 'string' },
      },
    },
    jwt: {
      type: 'object',
      required: ['algorithm'],
      properties: {
        algorithm: { type: 'string' },
      },
    },
    authentication: {
      type: 'object',
      required: ['secret', 'tokenExpiresIn', 'passwordLostExpiresIn'],
      properties: {
        secret: { type: 'string' },
        tokenExpiresIn: { type: 'string' },
        passwordLostExpiresIn: { type: 'string' },
      },
    },
    registration: {
      type: 'object',
      required: ['passwordSalt', 'secret', 'confirmationExpiresIn'],
      properties: {
        passwordSalt: { type: 'string' },
        secret: { type: 'string' },
        confirmationExpiresIn: { type: 'string' },
      },
    },
    smtp: {
      type: 'object',
      required: ['from', 'host', 'port'],
      additionalProperties: false,
      properties: {
        from: { type: 'string' },
        host: { type: 'string' },
        port: { type: 'number' },
        auth: {
          type: 'object',
          required: [],
          nullable: true,
          properties: {
            user: { type: 'string' },
            pass: { type: 'string' },
          },
        },
      },
    },
    datastore: {
      type: 'object',
      required: ['path'],
      properties: {
        path: { type: 'string' },
      },
    },
    development: {
      type: 'object',
      required: [],
      nullable: true,
      properties: {
        generateData: {
          type: 'object',
          nullable: true,
          required: ['users'],
          properties: {
            users: { type: 'number' },
            projectsPerUser: { type: 'number' },
          },
        },
        persistEmails: {
          type: 'boolean',
          nullable: true,
        },
      },
    },
    legalMentions: {
      type: 'string',
    },
    webapp: {
      type: 'object',
      nullable: true,
      properties: {
        appendToBody: {
          type: 'string',
          nullable: true,
        },
      },
    },
    userDocumentation: {
      type: 'object',
      nullable: true,
      properties: {
        appendToBody: {
          type: 'string',
          nullable: true,
        },
      },
    },
  },
};
