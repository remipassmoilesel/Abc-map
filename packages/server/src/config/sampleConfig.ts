/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 *
 *
 *
 */

import { ConfigInput } from './Config';

/**
 * This function return a valid javascript configuration example.
 */
export function sampleConfig(): string {
  const config: ConfigInput = {
    environmentName: 'local',
    externalUrl: 'http://localhost:10082/',
    server: {
      host: 'localhost',
      port: 10_082,
      log: {
        requests: true,
        errors: true,
        warnings: false,
      },
      globalRateLimit: {
        max: 1000,
        timeWindow: '1m',
      },
      authenticationRateLimit: {
        max: 1000,
        timeWindow: '1m',
      },
    },
    project: {
      maxPerUser: 10,
    },
    database: {
      url: 'mongodb://localhost:27019',
      username: 'admin',
      password: 'admin',
    },
    jwt: {
      algorithm: 'HS512',
    },
    authentication: {
      secret: 'azerty1234',
      tokenExpiresIn: '6h',
      passwordLostExpiresIn: '30min',
    },
    registration: {
      passwordSalt: 'azerty1234',
      secret: 'azerty1234',
      confirmationExpiresIn: '24h',
    },
    smtp: {
      from: 'no-reply@abc-map.fr',
      host: 'smtp.ethereal.email',
      port: 587,
      auth: {
        user: 'lelia16@ethereal.email',
        pass: '63rntn3G4DU3uue2MJ',
      },
    },
    datastore: {
      path: 'resources/dev-datastore',
    },
    development: {
      generateData: {
        users: 100,
        projectsPerUser: 5,
      },
      persistEmails: true,
    },
    webapp: {
      appendToBody: '<div>Server templated content (webapp)</div>',
    },
    userDocumentation: {
      appendToBody: '<div>Server templated content (userDocumentation)</div>',
    },
    legalMentions: `<div>This platform is offered in order to be useful, but ....</div>`,
  };

  return `
module.exports = ${JSON.stringify(config, null, 2)}
`;
}
