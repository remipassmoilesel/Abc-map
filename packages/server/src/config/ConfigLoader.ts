/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Env, EnvKey } from './Env';
import { Config, ConfigInput } from './Config';
import { errorMessage, Logger } from '@abc-map/shared';
import * as path from 'path';
import * as _ from 'lodash';
import { Validation } from '../utils/Validation';
import { isDirectory } from '../utils/isDirectory';
import { sampleConfig } from './sampleConfig';

export const logger = Logger.get('ConfigLoader.ts', 'info');

export type LoadConfigFunc = () => Promise<Config>;

export class ConfigLoader {
  public static readonly DEFAULT_CONFIG = 'resources/configuration/development.js';

  public static getPathFromEnv(): string {
    const env = new Env();
    return env.get(EnvKey.CONFIG) || ConfigLoader.DEFAULT_CONFIG;
  }

  public static async load(): Promise<Config> {
    const configPath = ConfigLoader.getPathFromEnv();
    return new ConfigLoader().load(configPath);
  }

  public static safeConfig(config: Config): Config {
    const safe: Config = _.cloneDeep(config);
    const safeValue = '<value-masked>';
    safe.authentication.secret = safeValue;
    safe.database.password = safeValue;
    safe.registration.passwordSalt = safeValue;
    safe.registration.secret = safeValue;
    if (safe.smtp.auth?.pass) {
      safe.smtp.auth.pass = safeValue;
    }

    safe.legalMentions = `${safe.legalMentions.substring(0, 50)}...`;

    return safe;
  }

  public async load(configPath: string): Promise<Config> {
    const _configPath = path.resolve(configPath);

    const explainedError = (message: string) => {
      logger.error(
        [
          '',
          '',
          '⚠️   Invalid configuration. ',
          '',
          'Message: ' + message,
          '',
          'This is a valid configuration example:',
          '',
          '------ config.js ------',
          sampleConfig()
            .split('\n')
            .map((line) => '   ' + line)
            .join('\n'),
          '------ config.js ------',
          '',
          'You can see other examples here: https://gitlab.com/abc-map/abc-map/-/tree/master/packages/server/resources/configuration',
          '',
          '',
        ].join('\n')
      );
    };

    let input: ConfigInput;
    try {
      /* eslint-disable @typescript-eslint/no-var-requires */
      input = _.cloneDeep(require(_configPath));
    } catch (err) {
      explainedError(errorMessage(err));
      return Promise.reject(new Error(`Cannot load configuration ${configPath}: ${errorMessage(err)}`));
    }

    // We validate config input
    if (!Validation.ConfigInput(input)) {
      explainedError(Validation.formatErrors(Validation.ConfigInput));
      return Promise.reject(new Error(`Configuration ${configPath} is not valid: ${Validation.formatErrors(Validation.ConfigInput)}`));
    }

    const config: Config = {
      ...input,
      webappPath: path.resolve(__dirname, '../../public/webapp'),
      userDocumentationPath: path.resolve(__dirname, '../../public/user-documentation'),
    };

    // We check if datastore is a directory
    if (!(await isDirectory(config.datastore.path))) {
      throw new Error(`Datastore root '${config.datastore.path}' must be a directory`);
    }

    // We remove trailing slash if present
    config.externalUrl = config.externalUrl.trim();
    if (config.externalUrl.slice(-1) === '/') {
      config.externalUrl = config.externalUrl.slice(0, -1);
    }

    // We check if webapp path is a directory
    if (!(await isDirectory(config.webappPath))) {
      throw new Error(`Webapp root '${config.webappPath}' must be a directory. Try "abc build" on a developer workstation.`);
    }

    return config;
  }
}
