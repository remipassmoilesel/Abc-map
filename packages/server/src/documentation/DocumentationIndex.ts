/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 *
 *
 *
 */

import * as util from 'util';
import * as path from 'path';
import { glob } from 'glob';
import { ConfigLoader, LoadConfigFunc } from '../config/ConfigLoader';
const globPromise = util.promisify(glob);

let instance: DocumentationIndex | undefined;

export class DocumentationIndex {
  public static get() {
    if (!instance) {
      instance = new DocumentationIndex(ConfigLoader.load);
    }
    return instance;
  }

  private cache: string[] = [];

  constructor(private config: LoadConfigFunc) {}

  /**
   * Returns the list of paths to documentation pages, relative to the documentation root
   */
  public async getEntries(): Promise<string[]> {
    if (!this.cache.length) {
      this.cache = await this.listEntries();
    }
    return this.cache;
  }

  private async listEntries(): Promise<string[]> {
    const { userDocumentationPath } = await this.config();

    const absolutePaths = await globPromise(userDocumentationPath + '/**/*.html', { nocase: true });
    return absolutePaths.map((absolutePath) => {
      return path.relative(userDocumentationPath, absolutePath);
    });
  }
}
