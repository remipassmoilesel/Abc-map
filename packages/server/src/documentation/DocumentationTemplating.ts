/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 *
 *
 *
 */

import { DocumentationIndex } from './DocumentationIndex';
import { promises as fs } from 'fs';
import { Config } from '../config/Config';
import * as path from 'path';

export const StartMark = '%%%append-to-body%%%';
export const EndMark = '%%%end-append-to-body%%%';

const replaceRegexp = new RegExp(StartMark + '[\\s\\S]*' + EndMark, 'm');

export class DocumentationTemplating {
  public static create(config: Config): DocumentationTemplating {
    return new DocumentationTemplating(config, DocumentationIndex.get());
  }

  constructor(private config: Config, private index: DocumentationIndex) {}

  public async templatePages() {
    const docRoot = this.config.userDocumentationPath;
    const toTemplate = this.config.userDocumentation?.appendToBody;
    if (!toTemplate) {
      return;
    }

    const pages = await this.index.getEntries();
    for (const pagePath of pages) {
      const absolutePath = path.resolve(docRoot, pagePath);
      const page = await fs.readFile(absolutePath);

      const originalContent = page.toString('utf8');
      const templated = originalContent.replace(replaceRegexp, StartMark + '  ' + toTemplate + '  ' + EndMark);
      if (originalContent !== templated) {
        await fs.writeFile(absolutePath, templated);
      }
    }
  }
}
