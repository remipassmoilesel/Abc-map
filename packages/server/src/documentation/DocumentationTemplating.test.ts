/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 *
 *
 *
 */

import { DocumentationTemplating, EndMark, StartMark } from './DocumentationTemplating';
import { Config, UserDocConfig } from '../config/Config';
import { SinonStubbedInstance } from 'sinon';
import { DocumentationIndex } from './DocumentationIndex';
import * as sinon from 'sinon';
import * as fs from 'fs';
import * as uuid from 'uuid-random';
import { expect } from 'chai';
import { execSync } from 'child_process';
import { ConfigLoader } from '../config/ConfigLoader';
import * as path from 'path';

describe('DocumentationTemplating', () => {
  describe('with a fake index', () => {
    let index: SinonStubbedInstance<DocumentationIndex>;
    let testRoot: string;
    let config: Partial<Config>;
    let templating: DocumentationTemplating;

    beforeEach(() => {
      index = sinon.createStubInstance(DocumentationIndex);

      testRoot = '/tmp/user-documentation-test-' + uuid();
      fs.mkdirSync(testRoot);

      config = {
        userDocumentation: {
          appendToBody: '<script src="/analytics.js"></script>\n<script src="/analytics-2.js"></script>',
        },
        userDocumentationPath: testRoot,
      };
      templating = new DocumentationTemplating(config as Config, index);
    });

    it('should work even if there are no page to template', async () => {
      index.getEntries.resolves([]);

      await templating.templatePages();
    });

    it('should template files if needed', async () => {
      // Prepare
      const index1 = `
      <html>
        <head><title>Documentation</title></head>
        <body>
            <h1>Documentation</h1>
            <div style="display: none">%%%append-to-body%%%   %%%end-append-to-body%%%</div>
        </body>
        </html>
    `;

      await fs.writeFileSync(testRoot + '/index-1.html', index1);
      await fs.writeFileSync(testRoot + '/index-2.html', '<html><head></head></html>\n\n\n');

      // Paths are relative to documentation root
      index.getEntries.resolves(['index-1.html', 'index-2.html']);

      // Act
      await templating.templatePages();

      // Assert
      expect(fs.readFileSync(testRoot + '/index-1.html').toString('utf-8')).toMatchSnapshot();
      expect(fs.readFileSync(testRoot + '/index-2.html').toString('utf-8')).toMatchSnapshot();
    });

    it('should template files twice if needed', async () => {
      // Prepare
      await fs.writeFileSync(
        testRoot + '/index-1.html',
        '<html><head></head><body><div style="display: none">%%%append-to-body%%%   %%%end-append-to-body%%%</div></body></html>'
      );
      await fs.writeFileSync(testRoot + '/index-2.html', '<html><head></head></html>\n\n\n');

      // Paths are relative to documentation root
      index.getEntries.resolves(['index-1.html', 'index-2.html']);

      await templating.templatePages();

      (config.userDocumentation as UserDocConfig).appendToBody =
        '<script src="/another-analytics.js"></script>\n<script src="/another-analytics-2.js"></script>';

      // Act
      await templating.templatePages();

      // Assert
      expect(fs.readFileSync(testRoot + '/index-1.html').toString('utf-8')).toMatchSnapshot();
      expect(fs.readFileSync(testRoot + '/index-2.html').toString('utf-8')).toMatchSnapshot();
    });
  });

  describe('with a real index', () => {
    let testRoot: string;
    let config: Config;
    let templating: DocumentationTemplating;

    beforeEach(async () => {
      config = await ConfigLoader.load();

      testRoot = '/tmp/user-documentation-test-' + uuid();
      fs.mkdirSync(testRoot);

      execSync(`cp -R ${config.userDocumentationPath}/* ${testRoot}`);

      templating = DocumentationTemplating.create({
        ...config,
        userDocumentation: {
          appendToBody: '<script src="/analytics.61ef55a258496.js"></script>\n<script src="/analytics.87fcfef6ee362852.js"></script>',
        },
        userDocumentationPath: testRoot,
      });
    });

    it('integration test', async () => {
      // Prepare
      const files = [path.resolve(testRoot + '/index.html'), path.resolve(testRoot + '/fr/index.html'), path.resolve(testRoot + '/en/index.html')];

      const beforeContent = files.map((filePath) => fs.readFileSync(filePath).toString('utf-8'));
      beforeContent.forEach((content) => {
        expect(content.includes(StartMark)).true;
        expect(content.includes(EndMark)).true;
      });

      // Act
      await templating.templatePages();

      // Assert
      const afterContent = files.map((filePath) => [filePath, fs.readFileSync(filePath).toString('utf-8')]);
      afterContent.forEach(([, content]) => {
        expect(content.includes(StartMark)).true;
        expect(content.includes(EndMark)).true;
        expect(content.includes('analytics.61ef55a258496.js')).true;
        expect(content.includes('analytics.87fcfef6ee362852.js')).true;
      });
    });
  });
});
