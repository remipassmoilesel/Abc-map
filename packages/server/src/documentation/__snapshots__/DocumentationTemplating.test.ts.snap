// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentationTemplating with a fake index should template files if needed 1`] = `
"
      <html>
        <head><title>Documentation</title></head>
        <body>
            <h1>Documentation</h1>
            <div style=\\"display: none\\">%%%append-to-body%%%  <script src=\\"/analytics.js\\"></script>
<script src=\\"/analytics-2.js\\"></script>  %%%end-append-to-body%%%</div>
        </body>
        </html>
    "
`;

exports[`DocumentationTemplating with a fake index should template files if needed 2`] = `
"<html><head></head></html>


"
`;

exports[`DocumentationTemplating with a fake index should template files twice if needed 1`] = `
"<html><head></head><body><div style=\\"display: none\\">%%%append-to-body%%%  <script src=\\"/another-analytics.js\\"></script>
<script src=\\"/another-analytics-2.js\\"></script>  %%%end-append-to-body%%%</div></body></html>"
`;

exports[`DocumentationTemplating with a fake index should template files twice if needed 2`] = `
"<html><head></head></html>


"
`;
